sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosDataWriteFactory.scala

```diff
@@ -25,27 +27,83 @@ class CosmosDataWriteFactory(userConfig: Map[String, String],
 
     val cosmosAccountConfig = CosmosAccountConfig.parseCosmosAccountConfig(userConfig)
     val cosmosTargetContainerConfig = CosmosContainerConfig.parseCosmosContainerConfig(userConfig)
+    val cosmosWriteConfig = CosmosWriteConfig.parseWriteConfig(userConfig)
 
     // TODO moderakh: this needs to be shared to avoid creating multiple clients
     val builder = new CosmosClientBuilder()
       .key(cosmosAccountConfig.key)
       .endpoint(cosmosAccountConfig.endpoint)
-      .consistencyLevel(ConsistencyLevel.EVENTUAL);
+      .consistencyLevel(ConsistencyLevel.EVENTUAL)
 
     SparkBridgeInternal.setMetadataCacheSnapshot(builder, cosmosClientStateHandle.value)
     val client = builder.buildAsyncClient();
 
+    val container = client.getDatabase(cosmosTargetContainerConfig.database)
+      .getContainer(cosmosTargetContainerConfig.container)
+
+    val containerDefinition = container.read().block().getProperties
+    val partitionKeyDefinition = containerDefinition.getPartitionKeyDefinition
+
     override def write(internalRow: InternalRow): Unit = {
-      // TODO moderakh: schema is hard coded for now to make end to end TestE2EMain work implement schema inference code
       val objectNode = CosmosRowConverter.fromInternalRowToObjectNode(internalRow, inputSchema)
-      // TODO: moderakh how should we handle absence of id?
-      if (!objectNode.has("id")) {
-        objectNode.put("id", UUID.randomUUID().toString)
+      ensureHasId(objectNode)
+
+      // TODO modreakh investigate if we should also support point write in non-blocking way
+      // TODO moderakh support patch?
```

Comment:
I think it would be good to have a quick call to go over how you see the write handshake to happen long-term - I am not quite sure which parts of the PR are just interim iterations or supposed to stay this way.
I haven't done exhaustive investigation into how other connectors are doing it - but from my understanding of reading through the Spark API/code I had the following protocol in mind:

write - would just put a ObjectNode based copy of the internal row into a Flux (with some reasonable buffer) as input - actual writes would happen via bulk operations. 
commit - would signal complete on the flux and block until processing has finished
abort - not sure - I don't think we have a way to rollback. When i glanced at other implementations (I think dynamodb and cassandra) they no-oped on abort. So this is what I had in mind (but the behavior might have changed in their connectors?)

My concern with blocking on writes is that this will reduce the achievable throughput significantly, won't it?

Again as long as we are more or less on the same page for the final design it is of course possible to do this in multiple iterations - just wanted to double-check whether we have the same final result in mind or not.

Reply:
Resolved in-person - Async will happen in subsequent PR